contracts/src/matching/HyperdriveMatchingEngineV2.sol

```diff
@@ -0,0 +1,659 @@
+///         the Hyperdrive AMM.
+/// @dev This version uses direct Hyperdrive mint/burn functions instead of flash 
+///      loans.
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+contract HyperdriveMatchingEngineV2 is 
+    IHyperdriveMatchingEngineV2,
+    ReentrancyGuard,
+    EIP712
+{
+    using FixedPointMath for uint256;
+    using SafeERC20 for ERC20;
+
+    /// @notice The EIP712 typehash of the OrderIntent struct.
+    bytes32 public constant ORDER_INTENT_TYPEHASH =
+        keccak256(
+            "OrderIntent(address trader,address counterparty,address hyperdrive,uint256 amount,uint256 slippageGuard,uint256 minVaultSharePrice,Options options,uint8 orderType,uint256 minMaturityTime,uint256 maxMaturityTime,uint256 expiry,bytes32 salt)"
+        );
+
+    /// @notice The EIP712 typehash of the Options struct
+    bytes32 public constant OPTIONS_TYPEHASH =
+        keccak256("Options(address destination,bool asBase)");
+
+    /// @notice The name of this matching engine
+    string public name;
+
+    /// @notice The kind of this matching engine
+    string public constant kind = HYPERDRIVE_MATCHING_ENGINE_KIND;
+
+    /// @notice The version of this matching engine
+    string public constant version = VERSION;
+
+    /// @notice The buffer amount used for cost related calculations.
+    uint256 public constant TOKEN_AMOUNT_BUFFER = 10;
+
+    /// @notice Mapping to track cancelled orders
+    mapping(bytes32 => bool) public isCancelled;
+
+    /// @notice Mapping to track the bond amount used for each order.
+    mapping(bytes32 => uint256) public orderBondAmountUsed;
+
+    /// @notice Mapping to track the amount of base used for each order.
+    mapping(bytes32 => uint256) public orderFundAmountUsed;
+
+    /// @notice Initializes the matching engine
+    /// @param _name The name of this matching engine
+    constructor(string memory _name) EIP712(_name, VERSION) {
+        name = _name;
+    }
+
+    /// @notice Matches two orders
+    /// @param _order1 The first order to match
+    /// @param _order2 The second order to match
+    /// @param _surplusRecipient The address that receives the surplus funds 
+    ///         from matching the trades
+    function matchOrders(
+        OrderIntent calldata _order1,
+        OrderIntent calldata _order2,
+        address _surplusRecipient
+    ) external nonReentrant {
+        // Validate orders
+        (bytes32 order1Hash, bytes32 order2Hash) = _validateOrders(
+            _order1, 
+            _order2
+        );
+
+        IHyperdrive hyperdrive = _order1.hyperdrive;
+
+        // Handle different order type combinations
+        if (_order1.orderType == OrderType.OpenLong && 
+            _order2.orderType == OrderType.OpenShort) {
+            // Case 1: Long + Short creation using mint()
```

Comment:
Let's put this comment above the if statement. We can add similar comments above the `else if` and `else` blocks.